zproject/backends.py

```diff
@@ -1470,6 +1483,133 @@ class GitLabAuthBackend(SocialAuthMixin, GitLabOAuth2):
     # we just use the primary email address, which is always verified.
     # (No code is required to do so, as that's the default behavior).
 
+@external_auth_method
+class AppleAuthBackend(SocialAuthMixin, AppleIdAuth):
+    sort_order = 10
+    name = "apple"
+    auth_backend_name = "Apple"
+    full_name_validated = True
+    REDIS_EXPIRATION_SECONDS = 60*10
+
+    @staticmethod
+    def get_apple_locale(django_language_code: str) -> str:
+        '''
+        Get the suitable apple supported locale with language code
+        for the Sign in / Continue with Apple buttons it provides.
+        '''
+        # The following is a list of locale values supported by Apple to send
+        # as params to URL which renders "Sign in with Apple" and "Continue with Apple"
+        # buttons. Gathered from
+        # https://developer.apple.com/documentation/signinwithapplejs/incorporating_sign_in_with_apple_into_other_platforms .
+        supported_locales = ['ar_SA', 'ca_ES', 'cs_CZ', 'da_DK', 'de_DE',
+                             'el_GR', 'en_US', 'es_ES', 'fi_FI', 'fr_FR',
+                             'hr_HR', 'hu_HU', 'id_ID', 'it_IT', 'iw_IL',
+                             'ja_JP', 'ko_KR', 'ms_MY', 'nl_NL', 'no_NO',
+                             'pl_PL', 'pt_PT', 'ro_RO', 'ru_RU', 'sk_SK',
+                             'sv_SE', 'th_TH', 'tr_TR', 'uk_UA', 'vi_VI',
+                             'zh_CN']
+        for locale in supported_locales:
+            if django_language_code in locale:
+                return locale
+        return 'en_US'
+
+    def get_or_create_state(self) -> str:
+        '''
+        Creates the Oauth2 state parameter in first step of the flow, before redirecting
+        the user to the IdP. We're dealing with Apple here, which will send the user back
+        to us with a POST request. Normally, we rely on being able to store certain parameters
+        in the user's session and use them after the redirect. Session is a Lax cookie, so
+        it won't be sent by the user's browser in the POST redirect from Apple. To get around this,
+        we will store the parameters in redis under a key derived from the state. That will allow us
+        to validate the state and retrieve the params after the redirect - by querying redis for
+        the key derived from the state sent in the POST redirect.
+        '''
+        request_data = self.strategy.request_data().dict()
+        data_to_store = {
+            key: request_data[key] for key in self.standard_relay_params
+            if key in request_data
+        }
+        state = self.state_token()
```

Comment:
Can you add a comment here explaining what format `state_token` has and what assumptions we can make about it (E.g. is it a unique random string generated by us, or is it under Apple or the user's control?)

Reply:
Added a comment. 
(It's a random string generated using PSA's internal function `state_token`)